File: C++/InterviewQuestions/Chess_GameDemo.cpp
/*
Chess Game
------------------------------

Functional Requirements:
    Two player chess game
    Initialize the board with pieces
    Piece movement strategy: every piece has its own rules
    Move a piece
    Validate a legal move
    Capture opponent pieces
    Detect Check
    Detect Checkmate
    Detect Stalemate

Pieces in a Chess Game:
    KING, QUEEN, PAWN, KNIGHT, BISHOP, ROOK

Entities:
    Player
    Game
    Board
    Move [Source Cell, Destination Cell, PieceMoved, capturedPiece]
    Piece [Abstract]: Color, PieceType, getPossibleMoves()
    King, Queen, Rook, Bishop, Knight, Pawn

Out of scope: castling, en passant, draw by repetition/50-move rule.
*/

#include <algorithm>
#include <iostream>
#include <memory>
#include <optional>
#include <stdexcept>
#include <string>
#include <vector>

using Position = std::pair<int, int>;  // (row, col), row 0 = rank 1 (white side), col 0 = file 'a'

enum class Color { WHITE, BLACK };

Color opposite(Color color) { return color == Color::WHITE ? Color::BLACK : Color::WHITE; }

enum class PieceType { KING, QUEEN, ROOK, BISHOP, KNIGHT, PAWN };

enum class GameStatus { ONGOING, CHECK, CHECKMATE, STALEMATE };

class Board;

class Piece {
public:
    Piece(Color color, PieceType type) : color_(color), type_(type) {}
    virtual ~Piece() = default;

    // Pseudo-legal moves: obeys movement/blocking/capture rules, ignores check safety.
    virtual std::vector<Position> getPossibleMoves(Position position, const Board& board) const = 0;

    // Squares this piece threatens; used for check detection. Same as possible
    // moves for every piece except the pawn, whose diagonal attacks aren't
    // conditional on an enemy being there yet.
    virtual std::vector<Position> getAttackedSquares(Position position, const Board& board) const {
        return getPossibleMoves(position, board);
    }

    char symbol() const {
        char letter;
        switch (type_) {
            case PieceType::KING: letter = 'K'; break;
            case PieceType::QUEEN: letter = 'Q'; break;
            case PieceType::ROOK: letter = 'R'; break;
            case PieceType::BISHOP: letter = 'B'; break;
            case PieceType::KNIGHT: letter = 'N'; break;
            case PieceType::PAWN: letter = 'P'; break;
        }
        return color_ == Color::WHITE ? letter : static_cast<char>(std::tolower(letter));
    }

    Color color() const { return color_; }
    PieceType type() const { return type_; }

protected:
    std::vector<Position> slide(Position position, const Board& board,
                                 const std::vector<Position>& directions) const;

    Color color_;
    PieceType type_;
};

class Board {
public:
    static constexpr int SIZE = 8;

    Board() : cells_(SIZE, std::vector<std::shared_ptr<Piece>>(SIZE, nullptr)) {}

    void initialize();

    bool isValidCell(Position position) const {
        auto [row, col] = position;
        return row >= 0 && row < SIZE && col >= 0 && col < SIZE;
    }

    std::shared_ptr<Piece> getPiece(Position position) const {
        auto [row, col] = position;
        return cells_[row][col];
    }

    void setPiece(Position position, std::shared_ptr<Piece> piece) {
        auto [row, col] = position;
        cells_[row][col] = std::move(piece);
    }

    std::shared_ptr<Piece> movePiece(Position source, Position destination) {
        auto captured = getPiece(destination);
        setPiece(destination, getPiece(source));
        setPiece(source, nullptr);
        return captured;
    }

    std::optional<Position> findKing(Color color) const {
        for (int row = 0; row < SIZE; ++row) {
            for (int col = 0; col < SIZE; ++col) {
                auto piece = cells_[row][col];
                if (piece && piece->type() == PieceType::KING && piece->color() == color) {
                    return Position{row, col};
                }
            }
        }
        return std::nullopt;
    }

    std::vector<std::pair<Position, std::shared_ptr<Piece>>> allPieces(Color color) const {
        std::vector<std::pair<Position, std::shared_ptr<Piece>>> result;
        for (int row = 0; row < SIZE; ++row) {
            for (int col = 0; col < SIZE; ++col) {
                auto piece = cells_[row][col];
                if (piece && piece->color() == color) {
                    result.push_back({Position{row, col}, piece});
                }
            }
        }
        return result;
    }

    void printBoard() const {
        for (int row = SIZE - 1; row >= 0; --row) {
            std::cout << (row + 1) << " ";
            for (int col = 0; col < SIZE; ++col) {
                auto piece = cells_[row][col];
                std::cout << (piece ? piece->symbol() : '.') << " ";
            }
            std::cout << "\n";
        }
        std::cout << "  a b c d e f g h\n";
    }

private:
    std::vector<std::vector<std::shared_ptr<Piece>>> cells_;
};

std::vector<Position> Piece::slide(Position position, const Board& board,
                                    const std::vector<Position>& directions) const {
    std::vector<Position> moves;
    auto [row, col] = position;
    for (auto [dRow, dCol] : directions) {
        int r = row + dRow;
        int c = col + dCol;
        while (board.isValidCell({r, c})) {
            auto occupant = board.getPiece({r, c});
            if (!occupant) {
                moves.push_back({r, c});
            } else {
                if (occupant->color() != color_) {
                    moves.push_back({r, c});
                }
                break;
            }
            r += dRow;
            c += dCol;
        }
    }
    return moves;
}

class Rook : public Piece {
public:
    explicit Rook(Color color) : Piece(color, PieceType::ROOK) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        return slide(position, board, {{1, 0}, {-1, 0}, {0, 1}, {0, -1}});
    }
};

class Bishop : public Piece {
public:
    explicit Bishop(Color color) : Piece(color, PieceType::BISHOP) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        return slide(position, board, {{1, 1}, {1, -1}, {-1, 1}, {-1, -1}});
    }
};

class Queen : public Piece {
public:
    explicit Queen(Color color) : Piece(color, PieceType::QUEEN) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        return slide(position, board,
                      {{1, 0}, {-1, 0}, {0, 1}, {0, -1}, {1, 1}, {1, -1}, {-1, 1}, {-1, -1}});
    }
};

class Knight : public Piece {
public:
    explicit Knight(Color color) : Piece(color, PieceType::KNIGHT) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        auto [row, col] = position;
        std::vector<Position> moves;
        static const std::vector<Position> offsets = {
            {2, 1}, {2, -1}, {-2, 1}, {-2, -1}, {1, 2}, {1, -2}, {-1, 2}, {-1, -2},
        };
        for (auto [dRow, dCol] : offsets) {
            Position candidate{row + dRow, col + dCol};
            if (board.isValidCell(candidate)) {
                auto occupant = board.getPiece(candidate);
                if (!occupant || occupant->color() != color_) {
                    moves.push_back(candidate);
                }
            }
        }
        return moves;
    }
};

class King : public Piece {
public:
    explicit King(Color color) : Piece(color, PieceType::KING) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        auto [row, col] = position;
        std::vector<Position> moves;
        for (int dRow = -1; dRow <= 1; ++dRow) {
            for (int dCol = -1; dCol <= 1; ++dCol) {
                if (dRow == 0 && dCol == 0) continue;
                Position candidate{row + dRow, col + dCol};
                if (board.isValidCell(candidate)) {
                    auto occupant = board.getPiece(candidate);
                    if (!occupant || occupant->color() != color_) {
                        moves.push_back(candidate);
                    }
                }
            }
        }
        return moves;
    }
};

class Pawn : public Piece {
public:
    explicit Pawn(Color color) : Piece(color, PieceType::PAWN) {}

    std::vector<Position> getPossibleMoves(Position position, const Board& board) const override {
        auto [row, col] = position;
        int direction = color_ == Color::WHITE ? 1 : -1;
        int startRow = color_ == Color::WHITE ? 1 : 6;
        std::vector<Position> moves;

        Position oneStep{row + direction, col};
        if (board.isValidCell(oneStep) && !board.getPiece(oneStep)) {
            moves.push_back(oneStep);
            Position twoStep{row + 2 * direction, col};
            if (row == startRow && !board.getPiece(twoStep)) {
                moves.push_back(twoStep);
            }
        }

        for (int dCol : {-1, 1}) {
            Position capture{row + direction, col + dCol};
            if (board.isValidCell(capture)) {
                auto occupant = board.getPiece(capture);
                if (occupant && occupant->color() != color_) {
                    moves.push_back(capture);
                }
            }
        }
        return moves;
    }

    std::vector<Position> getAttackedSquares(Position position, const Board& board) const override {
        auto [row, col] = position;
        int direction = color_ == Color::WHITE ? 1 : -1;
        std::vector<Position> squares;
        for (int dCol : {-1, 1}) {
            Position candidate{row + direction, col + dCol};
            if (board.isValidCell(candidate)) {
                squares.push_back(candidate);
            }
        }
        return squares;
    }
};

void Board::initialize() {
    static const std::vector<PieceType> backRank = {
        PieceType::ROOK, PieceType::KNIGHT, PieceType::BISHOP, PieceType::QUEEN,
        PieceType::KING, PieceType::BISHOP, PieceType::KNIGHT, PieceType::ROOK,
    };
    auto makePiece = [](PieceType type, Color color) -> std::shared_ptr<Piece> {
        switch (type) {
            case PieceType::ROOK: return std::make_shared<Rook>(color);
            case PieceType::KNIGHT: return std::make_shared<Knight>(color);
            case PieceType::BISHOP: return std::make_shared<Bishop>(color);
            case PieceType::QUEEN: return std::make_shared<Queen>(color);
            case PieceType::KING: return std::make_shared<King>(color);
            default: throw std::invalid_argument("Unsupported back-rank piece type");
        }
    };
    for (int col = 0; col < SIZE; ++col) {
        cells_[0][col] = makePiece(backRank[col], Color::WHITE);
        cells_[7][col] = makePiece(backRank[col], Color::BLACK);
    }
    for (int col = 0; col < SIZE; ++col) {
        cells_[1][col] = std::make_shared<Pawn>(Color::WHITE);
        cells_[6][col] = std::make_shared<Pawn>(Color::BLACK);
    }
}

struct Move {
    Position source;
    Position destination;
    std::shared_ptr<Piece> pieceMoved;
    std::shared_ptr<Piece> capturedPiece;
};

struct Player {
    std::string name;
    Color color;
};

Position parseSquare(const std::string& square) {
    int col = square[0] - 'a';
    int row = square[1] - '1';
    return {row, col};
}

class IllegalMoveError : public std::runtime_error {
public:
    explicit IllegalMoveError(const std::string& message) : std::runtime_error(message) {}
};

class Game {
public:
    Game(std::string whiteName = "White", std::string blackName = "Black")
        : white_{std::move(whiteName), Color::WHITE},
          black_{std::move(blackName), Color::BLACK},
          currentPlayer_(&white_),
          status_(GameStatus::ONGOING) {}

    void start() {
        board_ = Board();
        board_.initialize();
        currentPlayer_ = &white_;
        status_ = GameStatus::ONGOING;
        history_.clear();
    }

    void switchTurn() { currentPlayer_ = currentPlayer_ == &white_ ? &black_ : &white_; }

    bool isSquareAttacked(Position position, Color byColor) const {
        for (auto& [piecePosition, piece] : board_.allPieces(byColor)) {
            auto attacked = piece->getAttackedSquares(piecePosition, board_);
            if (std::find(attacked.begin(), attacked.end(), position) != attacked.end()) {
                return true;
            }
        }
        return false;
    }

    bool isCheck(Color color) const {
        auto kingPosition = board_.findKing(color);
        if (!kingPosition) return false;
        return isSquareAttacked(*kingPosition, opposite(color));
    }

    bool isCheckmate(Color color) { return isCheck(color) && !hasLegalMove(color); }

    bool isStalemate(Color color) { return !isCheck(color) && !hasLegalMove(color); }

    Move move(Position source, Position destination) {
        if (status_ == GameStatus::CHECKMATE || status_ == GameStatus::STALEMATE) {
            throw IllegalMoveError("Game is already over");
        }

        auto piece = board_.getPiece(source);
        if (!piece || piece->color() != currentPlayer_->color) {
            throw IllegalMoveError("No movable piece of the current player at source");
        }

        auto possible = piece->getPossibleMoves(source, board_);
        if (std::find(possible.begin(), possible.end(), destination) == possible.end()) {
            throw IllegalMoveError("Illegal move for this piece");
        }

        auto captured = board_.movePiece(source, destination);
        if (isCheck(currentPlayer_->color)) {
            // Undo: this move would leave (or keep) our own king in check.
            board_.setPiece(source, piece);
            board_.setPiece(destination, captured);
            throw IllegalMoveError("Move leaves own king in check");
        }

        if (piece->type() == PieceType::PAWN && (destination.first == 0 || destination.first == Board::SIZE - 1)) {
            piece = std::make_shared<Queen>(piece->color());
            board_.setPiece(destination, piece);
        }

        Move performedMove{source, destination, piece, captured};
        history_.push_back(performedMove);

        switchTurn();
        Color opponentColor = currentPlayer_->color;
        if (isCheckmate(opponentColor)) {
            status_ = GameStatus::CHECKMATE;
        } else if (isCheck(opponentColor)) {
            status_ = GameStatus::CHECK;
        } else if (isStalemate(opponentColor)) {
            status_ = GameStatus::STALEMATE;
        } else {
            status_ = GameStatus::ONGOING;
        }

        return performedMove;
    }

    const Player& currentPlayer() const { return *currentPlayer_; }
    GameStatus status() const { return status_; }
    Board& board() { return board_; }

    static std::string statusName(GameStatus status) {
        switch (status) {
            case GameStatus::ONGOING: return "ONGOING";
            case GameStatus::CHECK: return "CHECK";
            case GameStatus::CHECKMATE: return "CHECKMATE";
            case GameStatus::STALEMATE: return "STALEMATE";
        }
        return "UNKNOWN";
    }

private:
    bool hasLegalMove(Color color) {
        for (auto& [source, piece] : board_.allPieces(color)) {
            for (auto destination : piece->getPossibleMoves(source, board_)) {
                auto captured = board_.movePiece(source, destination);
                bool stillInCheck = isCheck(color);
                board_.setPiece(source, piece);
                board_.setPiece(destination, captured);
                if (!stillInCheck) return true;
            }
        }
        return false;
    }

    Board board_;
    Player white_;
    Player black_;
    Player* currentPlayer_;
    GameStatus status_;
    std::vector<Move> history_;
};

int main() {
    Game game("Alice", "Bob");
    game.start();

    // Fool's Mate -- fastest possible checkmate, four half-moves.
    std::vector<std::string> moves = {"f2f3", "e7e5", "g2g4", "d8h4"};
    for (const auto& moveStr : moves) {
        Position source = parseSquare(moveStr.substr(0, 2));
        Position destination = parseSquare(moveStr.substr(2, 2));
        std::string mover = game.currentPlayer().name;
        game.move(source, destination);
        std::cout << mover << " played " << moveStr << " -> status: " << Game::statusName(game.status()) << "\n";
    }

    std::cout << "\n";
    game.board().printBoard();
    std::cout << "\nFinal status: " << Game::statusName(game.status()) << "\n";

    return 0;
}

File: Python/InterviewQuestions/Chess_GameDemo.py
"""
Chess Game
------------------------------

Functional Requirements:
    Two player chess game
    Initialize the board with pieces
    Piece movement strategy: every piece has its own rules
    Move a piece
    Validate a legal move
    Capture opponent pieces
    Detect Check
    Detect Checkmate
    Detect Stalemate

Pieces in a Chess Game:
    KING, QUEEN, PAWN, KNIGHT, BISHOP, ROOK

Entities:
    Player
    Game
    Board
    Move [Source Cell, Destination Cell, PieceMoved, capturedPiece]
    Piece [Abstract]: Color, PieceType, get_possible_moves()
    King, Queen, Rook, Bishop, Knight, Pawn

Out of scope: castling, en passant, draw by repetition/50-move rule.
"""

from abc import ABC, abstractmethod
from enum import Enum, auto
from typing import List, Optional, Tuple

Position = Tuple[int, int]  # (row, col), row 0 = rank 1 (white side), col 0 = file 'a'


class Color(Enum):
    WHITE = auto()
    BLACK = auto()

    def opposite(self) -> "Color":
        return Color.BLACK if self == Color.WHITE else Color.WHITE


class PieceType(Enum):
    KING = auto()
    QUEEN = auto()
    ROOK = auto()
    BISHOP = auto()
    KNIGHT = auto()
    PAWN = auto()


class GameStatus(Enum):
    ONGOING = auto()
    CHECK = auto()
    CHECKMATE = auto()
    STALEMATE = auto()


class Piece(ABC):
    def __init__(self, color: Color, piece_type: PieceType):
        self.color = color
        self.type = piece_type

    @abstractmethod
    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        """Pseudo-legal moves: obeys movement/blocking/capture rules, ignores check safety."""
        raise NotImplementedError

    def get_attacked_squares(self, position: Position, board: "Board") -> List[Position]:
        """Squares this piece threatens; used for check detection. Same as possible
        moves for every piece except the pawn, whose diagonal attacks aren't
        conditional on an enemy being there yet."""
        return self.get_possible_moves(position, board)

    def symbol(self) -> str:
        letters = {
            PieceType.KING: "K", PieceType.QUEEN: "Q", PieceType.ROOK: "R",
            PieceType.BISHOP: "B", PieceType.KNIGHT: "N", PieceType.PAWN: "P",
        }
        letter = letters[self.type]
        return letter if self.color == Color.WHITE else letter.lower()

    def _slide(self, position: Position, board: "Board", directions: List[Position]) -> List[Position]:
        moves = []
        row, col = position
        for d_row, d_col in directions:
            r, c = row + d_row, col + d_col
            while board.is_valid_cell((r, c)):
                occupant = board.get_piece((r, c))
                if occupant is None:
                    moves.append((r, c))
                else:
                    if occupant.color != self.color:
                        moves.append((r, c))
                    break
                r, c = r + d_row, c + d_col
        return moves


class Rook(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.ROOK)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [(1, 0), (-1, 0), (0, 1), (0, -1)])


class Bishop(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.BISHOP)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [(1, 1), (1, -1), (-1, 1), (-1, -1)])


class Queen(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.QUEEN)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        return self._slide(position, board, [
            (1, 0), (-1, 0), (0, 1), (0, -1), (1, 1), (1, -1), (-1, 1), (-1, -1),
        ])


class Knight(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.KNIGHT)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        moves = []
        for d_row, d_col in [(2, 1), (2, -1), (-2, 1), (-2, -1), (1, 2), (1, -2), (-1, 2), (-1, -2)]:
            candidate = (row + d_row, col + d_col)
            if board.is_valid_cell(candidate):
                occupant = board.get_piece(candidate)
                if occupant is None or occupant.color != self.color:
                    moves.append(candidate)
        return moves


class King(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.KING)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        moves = []
        for d_row in (-1, 0, 1):
            for d_col in (-1, 0, 1):
                if d_row == 0 and d_col == 0:
                    continue
                candidate = (row + d_row, col + d_col)
                if board.is_valid_cell(candidate):
                    occupant = board.get_piece(candidate)
                    if occupant is None or occupant.color != self.color:
                        moves.append(candidate)
        return moves


class Pawn(Piece):
    def __init__(self, color: Color):
        super().__init__(color, PieceType.PAWN)

    def get_possible_moves(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        direction = 1 if self.color == Color.WHITE else -1
        start_row = 1 if self.color == Color.WHITE else 6
        moves = []

        one_step = (row + direction, col)
        if board.is_valid_cell(one_step) and board.get_piece(one_step) is None:
            moves.append(one_step)
            two_step = (row + 2 * direction, col)
            if row == start_row and board.get_piece(two_step) is None:
                moves.append(two_step)

        for d_col in (-1, 1):
            capture = (row + direction, col + d_col)
            if board.is_valid_cell(capture):
                occupant = board.get_piece(capture)
                if occupant is not None and occupant.color != self.color:
                    moves.append(capture)
        return moves

    def get_attacked_squares(self, position: Position, board: "Board") -> List[Position]:
        row, col = position
        direction = 1 if self.color == Color.WHITE else -1
        squares = []
        for d_col in (-1, 1):
            candidate = (row + direction, col + d_col)
            if board.is_valid_cell(candidate):
                squares.append(candidate)
        return squares


class Move:
    def __init__(self, source: Position, destination: Position,
                 piece_moved: Piece, captured_piece: Optional[Piece]):
        self.source = source
        self.destination = destination
        self.piece_moved = piece_moved
        self.captured_piece = captured_piece


class Board:
    SIZE = 8

    def __init__(self):
        self.cells: List[List[Optional[Piece]]] = [[None] * Board.SIZE for _ in range(Board.SIZE)]
        self.initialize()

    def initialize(self) -> None:
        back_rank = [PieceType.ROOK, PieceType.KNIGHT, PieceType.BISHOP, PieceType.QUEEN,
                     PieceType.KING, PieceType.BISHOP, PieceType.KNIGHT, PieceType.ROOK]
        piece_classes = {
            PieceType.ROOK: Rook, PieceType.KNIGHT: Knight, PieceType.BISHOP: Bishop,
            PieceType.QUEEN: Queen, PieceType.KING: King,
        }
        for col, piece_type in enumerate(back_rank):
            self.cells[0][col] = piece_classes[piece_type](Color.WHITE)
            self.cells[7][col] = piece_classes[piece_type](Color.BLACK)
        for col in range(Board.SIZE):
            self.cells[1][col] = Pawn(Color.WHITE)
            self.cells[6][col] = Pawn(Color.BLACK)

    def is_valid_cell(self, position: Position) -> bool:
        row, col = position
        return 0 <= row < Board.SIZE and 0 <= col < Board.SIZE

    def get_piece(self, position: Position) -> Optional[Piece]:
        row, col = position
        return self.cells[row][col]

    def set_piece(self, position: Position, piece: Optional[Piece]) -> None:
        row, col = position
        self.cells[row][col] = piece

    def move_piece(self, source: Position, destination: Position) -> Optional[Piece]:
        captured = self.get_piece(destination)
        self.set_piece(destination, self.get_piece(source))
        self.set_piece(source, None)
        return captured

    def find_king(self, color: Color) -> Optional[Position]:
        for row in range(Board.SIZE):
            for col in range(Board.SIZE):
                piece = self.cells[row][col]
                if piece is not None and piece.type == PieceType.KING and piece.color == color:
                    return (row, col)
        return None

    def all_pieces(self, color: Color) -> List[Tuple[Position, Piece]]:
        result = []
        for row in range(Board.SIZE):
            for col in range(Board.SIZE):
                piece = self.cells[row][col]
                if piece is not None and piece.color == color:
                    result.append(((row, col), piece))
        return result

    def print_board(self) -> None:
        for row in range(Board.SIZE - 1, -1, -1):
            rank = [self.cells[row][col].symbol() if self.cells[row][col] else "." for col in range(Board.SIZE)]
            print(row + 1, " ".join(rank))
        print("  " + " ".join("abcdefgh"))


class Player:
    def __init__(self, name: str, color: Color):
        self.name = name
        self.color = color


def parse_square(square: str) -> Position:
    col = ord(square[0]) - ord("a")
    row = int(square[1]) - 1
    return (row, col)


class IllegalMoveError(Exception):
    pass


class Game:
    def __init__(self, white_name: str = "White", black_name: str = "Black"):
        self.board = Board()
        self.white = Player(white_name, Color.WHITE)
        self.black = Player(black_name, Color.BLACK)
        self.current_player = self.white
        self.status = GameStatus.ONGOING
        self.history: List[Move] = []

    def start(self) -> None:
        self.board.initialize()
        self.current_player = self.white
        self.status = GameStatus.ONGOING
        self.history.clear()

    def switch_turn(self) -> None:
        self.current_player = self.black if self.current_player is self.white else self.white

    def is_square_attacked(self, position: Position, by_color: Color) -> bool:
        for piece_position, piece in self.board.all_pieces(by_color):
            if position in piece.get_attacked_squares(piece_position, self.board):
                return True
        return False

    def is_check(self, color: Color) -> bool:
        king_position = self.board.find_king(color)
        if king_position is None:
            return False
        return self.is_square_attacked(king_position, color.opposite())

    def _has_legal_move(self, color: Color) -> bool:
        for source, piece in self.board.all_pieces(color):
            for destination in piece.get_possible_moves(source, self.board):
                captured = self.board.move_piece(source, destination)
                still_in_check = self.is_check(color)
                self.board.set_piece(source, piece)
                self.board.set_piece(destination, captured)
                if not still_in_check:
                    return True
        return False

    def is_checkmate(self, color: Color) -> bool:
        return self.is_check(color) and not self._has_legal_move(color)

    def is_stalemate(self, color: Color) -> bool:
        return not self.is_check(color) and not self._has_legal_move(color)

    def move(self, source: Position, destination: Position) -> Move:
        if self.status in (GameStatus.CHECKMATE, GameStatus.STALEMATE):
            raise IllegalMoveError("Game is already over")

        piece = self.board.get_piece(source)
        if piece is None or piece.color != self.current_player.color:
            raise IllegalMoveError("No movable piece of the current player at source")

        if destination not in piece.get_possible_moves(source, self.board):
            raise IllegalMoveError("Illegal move for this piece")

        captured = self.board.move_piece(source, destination)
        if self.is_check(self.current_player.color):
            # Undo: this move would leave (or keep) our own king in check.
            self.board.set_piece(source, piece)
            self.board.set_piece(destination, captured)
            raise IllegalMoveError("Move leaves own king in check")

        if piece.type == PieceType.PAWN and destination[0] in (0, Board.SIZE - 1):
            piece = Queen(piece.color)
            self.board.set_piece(destination, piece)

        move = Move(source, destination, piece, captured)
        self.history.append(move)

        self.switch_turn()
        opponent_color = self.current_player.color
        if self.is_checkmate(opponent_color):
            self.status = GameStatus.CHECKMATE
        elif self.is_check(opponent_color):
            self.status = GameStatus.CHECK
        elif self.is_stalemate(opponent_color):
            self.status = GameStatus.STALEMATE
        else:
            self.status = GameStatus.ONGOING

        return move


def main() -> None:
    game = Game("Alice", "Bob")
    game.start()

    # Fool's Mate -- fastest possible checkmate, four half-moves.
    moves = ["f2f3", "e7e5", "g2g4", "d8h4"]
    for move_str in moves:
        source = parse_square(move_str[0:2])
        destination = parse_square(move_str[2:4])
        mover = game.current_player.name
        game.move(source, destination)
        print(f"{mover} played {move_str} -> status: {game.status.name}")

    print()
    game.board.print_board()
    print(f"\nFinal status: {game.status.name}")


if __name__ == "__main__":
    main()
